// dawg-db/index.js
/*

    DAWGDB

    (C) DAWGCLUB
    
*/

const fs = require('fs');
const { 
    v1: uuidv1,
    v4: uuidv4,
} = require('uuid');

exports.Database = class {
    name;
    items;

    constructor(name, key=uuidv4()) {
        this.name = name;
        this.items = [];

        if(this.isInitialized()) {
            this.items = this.get();
        }

        this.update()
    }

    add(item) {
        this.items.push(item);
        this.update();
        return item;
    }

    remove(item) {
        this.items.splice(this.items.indexOf(item), 1);
        this.update();
        return item;
    }


    get() {
        if(this.isInitialized()) {
            this.items = JSON.parse(fs.readFileSync('./db/' + this.name + '.json'));
            return JSON.parse(fs.readFileSync('./db/' + this.name + '.json'));
        }
        
        return [];
    }

    query(query) {
        if(!this.isInitialized()) return null;

        this.get();

        const queries = Object.getOwnPropertyNames(query)
        let results = []

        for(let i = 0; i < queries.length; i++) {
            for (let item of this.items) {
                if(item[queries[0]] === query[queries[0]]) {
                    results.push(item);
                }
                else {
                    if(results.includes(item)) results.splice(results.indexOf(item), 1);
                }
            }
        }


        if(results.length === 1) return results[0];
        if(results.length < 1) return null;

        return results;
    }

    update() {
        if(!this.isGlobeInitialized()) fs.mkdirSync('./db/');

        fs.writeFileSync('./db/' + this.name + '.json', JSON.stringify(this.items));
    }

    isGlobeInitialized() {
        try {
            // check if variable
            return fs.existsSync('./db/');
        }
        catch(e) {
            return false;
        }
    }

    isDatabaseInitialized() {
        return fs.existsSync('./db/' + this.name + '.json');
    }

    isInitialized() {
        return this.isDatabaseInitialized() && this.isGlobeInitialized();
    }

    delete() {
        if(!this.isGlobeInitialized()) return;

        fs.unlinkSync('./db/' + this.name + '.json');
    }

    dashboard(port=3000) {
        // express
        const express = require('express');

        const app = express();

        app.get('/', (req, res) => {
            // read dashboard.html content
            var dashboardHTML = fs.readFileSync('./lib/dashboard.html', 'utf8');

            // replace placeholders
            dashboardHTML = dashboardHTML
                .replace('${db_name}', this.name)
                .replace('${db_json}', JSON.stringify(this.items).replace("'", "\\'"));

            res.send(dashboardHTML)
            
        });

        app.get('/set', (req, res) => {
            // read get request
            const data = req.query;

            this.items[data.obj][data.prop] = data.val;
            this.update()
        })

        app.listen(port, () => console.log('DawgDB dashboard running on port ' + port));
    }
}
